packages/react-client/src/ReactFlightReplyClient.js

```diff
@@ -1151,64 +1154,86 @@ export function registerBoundServerReference<T: Function>(
     Object.defineProperties((reference: any), {
       $$FORM_ACTION: {value: $$FORM_ACTION},
       $$IS_SIGNATURE_EQUAL: {value: isSignatureEqual},
-      bind: {value: bind},
     });
+    defineBind(reference);
   }
   knownServerReferences.set(reference, {id, bound});
 }
 
+// TODO: Ideally we'd use `isServerReference` from
+// 'react-server/src/ReactFlightServerConfig', but that can only be imported
+// in a react-server environment.
+function isServerReference(reference: Object): boolean {
+  return reference.$$typeof === Symbol.for('react.server.reference');
+}
+
 export function registerServerReference<T: Function>(
   reference: T,
   id: ServerReferenceId,
   encodeFormAction?: EncodeFormActionCallback,
 ): ServerReference<T> {
-  registerBoundServerReference(reference, id, null, encodeFormAction);
+  const bound =
+    isServerReference(reference) && reference.$$bound
+      ? Promise.resolve(reference.$$bound)
+      : null;
+
+  registerBoundServerReference(reference, id, bound, encodeFormAction);
   return reference;
 }
 
 // $FlowFixMe[method-unbinding]
 const FunctionBind = Function.prototype.bind;
 // $FlowFixMe[method-unbinding]
 const ArraySlice = Array.prototype.slice;
-function bind(this: Function): Function {
-  // $FlowFixMe[unsupported-syntax]
-  // $FlowFixMe[prop-missing]
-  const newFn = FunctionBind.apply(this, arguments);
-  const reference = knownServerReferences.get(this);
-  if (reference) {
-    if (__DEV__) {
-      const thisBind = arguments[0];
-      if (thisBind != null) {
-        // This doesn't warn in browser environments since it's not instrumented outside
-        // usedWithSSR. This makes this an SSR only warning which we don't generally do.
-        // TODO: Consider a DEV only instrumentation in the browser.
-        console.error(
-          'Cannot bind "this" of a Server Action. Pass null or undefined as the first argument to .bind().',
+
+function defineBind<T: Function>(reference: T) {
+  // TODO: Instead of checking if it's a server reference, we could also use
+  // `reference.hasOwnProperty('bind')`.
+  const originalBind = isServerReference(reference)
```

Comment:
We don't even have to check. We can just assume that it has a `bind` since it's a function. Otherwise something has gone terribly wrong and if not we probably shouldn't add one.

The principle here is this should be opaque to the environment and as long as we just call through it should work as usual.

In fact, you should really be able to have the same function in multiple server environments too which would be possible if we moved the expandos to WeakMaps.

So this could be bound for the Flight Client, Flight Server ESM and Flight Webpack all at once.